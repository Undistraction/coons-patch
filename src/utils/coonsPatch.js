import {
  interpolateStraightLineOnXAxis,
  interpolateStraightLineOnYAxis,
} from './interpolate/curves/straight'
import { interpolatePointOnCurveEvenlySpaced } from './interpolate/pointOnCurve/even'
import { interpolatePointOnSurface } from './interpolate/pointOnSurface/bilinear'

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

const addAll = (list) => list.reduce((total, { value }) => total + value, 0)

const getCounts = (columns, rows) => ({
  columnsTotalCount: columns.length,
  rowsTotalCount: rows.length,
})

const getTotalValues = (columns, rows) => ({
  columnsTotalValue: addAll(columns),
  rowsTotalValue: addAll(rows),
})

const getTSize = (steps, idx, totalValue) => {
  const step = steps[idx]
  const stepValue = step.value || step
  return stepValue / totalValue
}

const getStepData = (columns, rows) => ({
  ...getCounts(columns, rows),
  ...getTotalValues(columns, rows),
})

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

export const getCurvesOnXAxis = (
  boundingCurves,
  columns,
  rows,
  interpolateLineOnXAxis = interpolateStraightLineOnXAxis,
  interpolatePointOnCurve = interpolatePointOnCurveEvenlySpaced
) => {
  const {
    columnsTotalCount,
    rowsTotalCount,
    columnsTotalValue,
    rowsTotalValue,
  } = getStepData(columns, rows)

  const curves = []
  let vStart = 0

  // Short circuit if we are only 1x1 and just return bounds
  if (columnsTotalCount === 1 && rowsTotalCount === 1) {
    return [[boundingCurves.top], [boundingCurves.bottom]]
  }

  for (let rowIdx = 0; rowIdx <= rowsTotalCount; rowIdx++) {
    const lineSections = []
    let uStart = 0

    for (let columnIdx = 0; columnIdx < columnsTotalCount; columnIdx++) {
      const column = columns[columnIdx]
      const columnValue = column?.value
      const uSize = columnValue / columnsTotalValue
      const uEnd = uStart + uSize

      if (!column.isGutter) {
        const curve = interpolateLineOnXAxis(
          boundingCurves,
          uStart,
          uSize,
          uEnd,
          vStart,
          interpolatePointOnCurve
        )

        lineSections.push(curve)
      }

      uStart += uSize
    }

    curves.push(lineSections)

    if (rowIdx !== rowsTotalCount) {
      vStart += getTSize(rows, rowIdx, rowsTotalValue)
    }
  }

  return curves
}

export const getCurvesOnYAxis = (
  boundingCurves,
  columns,
  rows,
  interpolateLineOnYAxis = interpolateStraightLineOnYAxis,
  interpolatePointOnCurve = interpolatePointOnCurveEvenlySpaced
) => {
  const {
    columnsTotalCount,
    rowsTotalCount,
    columnsTotalValue,
    rowsTotalValue,
  } = getStepData(columns, rows)

  const curves = []
  let uStart = 0

  // Short circuit if we are only 1x1 and just return the bounds
  if (columnsTotalCount === 1 && rowsTotalCount === 1) {
    return [[boundingCurves.left], [boundingCurves.right]]
  }

  for (let columnIdx = 0; columnIdx <= columnsTotalCount; columnIdx++) {
    const lineSections = []
    let vStart = 0

    for (let rowIdx = 0; rowIdx < rowsTotalCount; rowIdx++) {
      const row = rows[rowIdx]
      const rowValue = row?.value
      const vSize = rowValue / rowsTotalValue
      const vEnd = vStart + vSize

      if (!row.isGutter) {
        const curve = interpolateLineOnYAxis(
          boundingCurves,
          vStart,
          vSize,
          vEnd,
          uStart,
          interpolatePointOnCurve
        )

        lineSections.push(curve)
      }

      vStart += vSize
    }

    curves.push(lineSections)

    // Calculate the position of the next column
    if (columnIdx !== columnsTotalCount) {
      uStart += getTSize(columns, columnIdx, columnsTotalValue)
    }
  }

  return curves
}

export const getCurveIntersections = (
  boundingCurves,
  columns,
  rows,
  interpolatePointOnCurve
) => {
  const {
    columnsTotalCount,
    rowsTotalCount,
    columnsTotalValue,
    rowsTotalValue,
  } = getStepData(columns, rows)

  const intersections = []
  let vStart = 0

  for (let rowIdx = 0; rowIdx <= rowsTotalCount; rowIdx++) {
    let uStart = 0

    for (let columnIdx = 0; columnIdx <= columnsTotalCount; columnIdx++) {
      const point = interpolatePointOnSurface(
        boundingCurves,
        uStart,
        vStart,
        interpolatePointOnCurve
      )

      intersections.push(point)

      if (columnIdx !== columnsTotalCount) {
        uStart += getTSize(columns, columnIdx, columnsTotalValue)
      }
    }

    if (rowIdx !== rowsTotalCount) {
      vStart += getTSize(rows, rowIdx, rowsTotalValue)
    }
  }

  return intersections
}
