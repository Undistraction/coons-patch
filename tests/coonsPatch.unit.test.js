import coonsPatch from '../src'
import fixtures, { boundingCurvesValid, paramsObjValid } from './fixtures'
import testValidationOfBoundingCurveArgs from './shared/testValidationOfBoundingCurveArgs'

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

describe.skip(`coonsPatch`, () => {
  describe(`validations`, () => {
    testValidationOfBoundingCurveArgs((boundingCurves) =>
      coonsPatch(boundingCurves, 0, 1)
    )

    describe(`params object`, () => {
      describe.each([
        {
          key: `u`,
        },
        {
          key: `v`,
        },
      ])(`key: $key`, ({ key }) => {
        it(`should throw if ${key} is not a number`, () => {
          expect(() =>
            coonsPatch(boundingCurvesValid, { ...paramsObjValid, [key]: `abc` })
          ).toThrow(`params.${key} value must be a number, but was 'abc'`)
        })

        it(`should throw if ${key} is less than zero`, () => {
          expect(() =>
            coonsPatch(boundingCurvesValid, { ...paramsObjValid, [key]: -0.1 })
          ).toThrow(
            `params.${key} value must be between 0 and 1, but was '-0.1'`
          )
        })

        it(`should throw if ${key} is greater than 1`, () => {
          expect(() =>
            coonsPatch(boundingCurvesValid, { ...paramsObjValid, [key]: 1.1 })
          ).toThrow(
            `params.${key} value must be between 0 and 1, but was '1.1'`
          )
        })
      })
    })

    describe(`interpolatePointOnCurveU`, () => {
      it(`should throw if interpolatePointOnCurveU is not a function`, () => {
        expect(() =>
          coonsPatch(boundingCurvesValid, paramsObjValid, {
            interpolatePointOnCurveU: 123,
          })
        ).toThrow(`interpolatePointOnCurveU must be a function`)
      })
    })

    describe(`interpolatePointOnCurveV`, () => {
      it(`should throw if interpolatePointOnCurveV is not a function`, () => {
        expect(() =>
          coonsPatch(boundingCurvesValid, paramsObjValid, {
            interpolatePointOnCurveV: 123,
          })
        ).toThrow(`interpolatePointOnCurveV must be a function`)
      })
    })
  })

  describe(`success`, () => {
    it(`doesn't throw with minimal params`, () => {
      expect(() =>
        coonsPatch(boundingCurvesValid, paramsObjValid)
      ).not.toThrow()
    })

    describe(`corner points`, () => {
      it(`returns 0/0 for top-left corner`, () => {
        expect(coonsPatch(boundingCurvesValid, { u: 0, v: 0 })).toEqual({
          x: 0,
          y: 0,
        })
      })

      it(`returns width/0 for top-right corner`, () => {
        expect(coonsPatch(boundingCurvesValid, { u: 1, v: 0 })).toEqual({
          x: 100,
          y: 0,
        })
      })

      it(`returns width/0 for bottom-left corner`, () => {
        expect(coonsPatch(boundingCurvesValid, { u: 0, v: 1 })).toEqual({
          x: 0,
          y: 100,
        })
      })

      it(`returns width/height for bottom-right corner`, () => {
        expect(coonsPatch(boundingCurvesValid, { u: 1, v: 1 })).toEqual({
          x: 100,
          y: 100,
        })
      })
    })

    describe.each(fixtures)(`fixture: '$name'`, ({ input }) => {
      it(`returns the correct data`, () => {
        expect(coonsPatch(...input.coonsPatch.args)).toMatchSnapshot()
      })
    })
  })
})
