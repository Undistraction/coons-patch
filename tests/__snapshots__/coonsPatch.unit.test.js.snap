// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`coonsPatch > success > fixture: ''Arbitrary point with different interpâ€¦'' > returns the correct data 1`] = `
{
  "x": 11.875,
  "y": 50,
}
`;

exports[`coonsPatch > success > fixture: ''Arbitrary point with linear interpolaâ€¦'' > returns the correct data 1`] = `
{
  "x": 11.875,
  "y": 50,
}
`;

exports[`coonsPatch > success > fixture: ''Arbitrary point'' > returns the correct data 1`] = `
{
  "x": 20.174815736217084,
  "y": 50.15502250526892,
}
`;
